lib/rubygems/resolver.rb

```diff
@@ -233,8 +233,29 @@ def search_for(dependency)
       exc.errors = @set.errors
       raise exc
     end
-    possibles.sort_by { |s| [s.source, s.version, Gem::Platform.local =~ s.platform ? 1 : 0] }.
-      map { |s| ActivationRequest.new s, dependency, [] }
+
+    sources = []
+    groups = Hash.new { |hash, key| hash[key] = [] }
+
+    possibles.each do |spec|
+      source = spec.source
+
+      sources << source unless sources.include? source
+
+      groups[source] << spec
+    end
+
+    sources.sort.each_with_object [] do |source, activation_requests|
+      specs = groups[source]
+
+      specs.sort_by! { |spec| [spec.version, Gem::Platform.local =~ spec.platform ? 1 : 0] }
+
+      specs.each do |spec|
+        activation_request = ActivationRequest.new spec, dependency, []
+
+        activation_requests << activation_request
+      end
+    end
```

Comment:
Unfortunately, the source ordering has to be taken into account before sorting the versions and platform locality. The original sort ordering supplied by the user is not available in this lexical context. We can only rely on the assumption that the gems were returned from `#find_possible` in order of their source. This assumption holds because the fetcher iterates over each source in order to retrieve the specs.